From 6bc835e424eddefe15fc1d35a5b33230168f80e1 Mon Sep 17 00:00:00 2001
From: Christian Brauner <brauner@kernel.org>
Date: Thu, 23 Apr 2026 11:56:04 +0200
Subject: [PATCH] eventpoll: use hlist_is_singular_node() in __ep_remove()

jira VULN-186882
cve-pre CVE-2026-46242
commit-author Christian Brauner <brauner@kernel.org>
commit 3d9fd0abc94d8cd430cc7cd7d37ce5e5aae2cd2b

Replace the open-coded "epi is the only entry in file->f_ep" check
with hlist_is_singular_node(). Same semantics, and the helper avoids
the head-cacheline access in the common false case.

Link: https://patch.msgid.link/20260423-work-epoll-uaf-v1-1-2470f9eec0f5@kernel.org
Signed-off-by: Christian Brauner (Amutable) <brauner@kernel.org>

(cherry picked from commit 3d9fd0abc94d8cd430cc7cd7d37ce5e5aae2cd2b)
Signed-off-by: Shreeya Patel <spatel@ciq.com>
---
 fs/eventpoll.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/fs/eventpoll.c b/fs/eventpoll.c
index c2b9630e9883..fbf4345581c0 100644
--- a/fs/eventpoll.c
+++ b/fs/eventpoll.c
@@ -741,7 +741,7 @@ static bool __ep_remove(struct eventpoll *ep, struct epitem *epi, bool force)
 
 	to_free = NULL;
 	head = file->f_ep;
-	if (head->first == &epi->fllink && !epi->fllink.next) {
+	if (hlist_is_singular_node(&epi->fllink, head)) {
 		/* See eventpoll_release() for details. */
 		WRITE_ONCE(file->f_ep, NULL);
 		if (!is_file_epoll(file)) {
-- 
2.52.0

